### DISCUSSION TITLE
support of gluster-kadalu on windows node on kubernetes

### DISCUSSION THREAD

**Discussion by themonaminor** [kadalu/kadalu#518] [Feature Requests]
working on k8s cluster where linux and windows nodes are present.
since gluster-kadalu is working on linux only node. so storing or consuming storage from gluster for windows nodes in k8s is seems to be difficult.
wanted some inputs in this aspect.

**Comment by amarts** [comment]
Well, I am not much aware of Windows node part of k8s, specially if it allows `csi-nodeplugin` to work or not.

The requirement to get a kadalu PV on any node, we need below:

* a node which supports running kadalu's `csi-nodeplugin`.
* a node which can then support fuse and mounting glusterfs on the host.

Is there any other linux storage specific projects work on windows node? Now, another way to look at it is, if we get blockStorage support on windows, would this work?

Any major pointers to this would be welcome! We are happy to look into this and add it into our roadmap if this is feasible.

**Comment by hunter86bg** [comment]
Recently I have been asked to deploy a storage solution that works for both Linux and Windows + Kubernetes. Naturally, GlusterFS is a good candidate.
So far, I've found out that only kadalu fits my needs yet I have to use it in 'external' mode where a regular gluster volume will be presented via Samba to Windows, FUSE for linux and kadalu for Kubernetes-native storage support

**Comment by leelavg** [comment]
Leaving it here for future reference https://github.com/kubernetes-csi/csi-proxy

**Reply by amarts** [reply]
Thanks for this reference. Thinking more on this, I guess we can just utilize the [SMB Driver project](https://github.com/kubernetes-csi/csi-driver-smb/blob/master/pkg/mounter/safe_mounter_windows.go) linked above as client on windows, and we can build a gluster smb exporter application.

**Reply by leelavg** [reply]
@amarts since we'll be have repl vols as well, I think we need CTDB and all the VIPs embargo. Or following this [doc](https://docs.gluster.org/en/latest/Administrator-Guide/Setting-Up-Clients/#exporting-gluster-volumes-through-samba) is good enough?

**Comment by hunter86bg** [comment]
Usually, it's alittle bit more complicated.
We need to:

- install samba with libgfapi modules
- open the firewall
- set user.cifs option on the volume
- set samba group of settings
~~~
gluster volume set { gluster_cluster_volume } group samba
~~~
- trigger the samba hook, so we don't stop+start the volume:
~~~
/var/lib/glusterd/hooks/1/start/post/S30samba-start.sh --gd-workdir /var/lib/glusterd --volname { gluster_cluster_volume }
~~~
- Set user that will have access to the share (uid should be the same everywhere or the mount point should be 777)
-  Mount the fuse mount point with acl option and setfacl -m u:<samba_user>:rwx or rely on 777 and controll access from samba
 
CTDB won't be needed if we use a container for the share
